File: backend/src/main/java/ma/investpro/entity/Fournisseur.java
package ma.investpro.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;

/**
 * Entité représentant un fournisseur
 */
@Entity
@Table(name = "fournisseurs", uniqueConstraints = {
    @UniqueConstraint(columnNames = "code"),
    @UniqueConstraint(columnNames = "ice")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Fournisseur extends BaseEntity {

    @Column(nullable = false, unique = true, length = 50)
    @NotBlank(message = "Le code du fournisseur est obligatoire")
    @Size(max = 50, message = "Le code ne peut pas dépasser 50 caractères")
    private String code;

    @Column(nullable = false, length = 200)
    @NotBlank(message = "La raison sociale est obligatoire")
    @Size(max = 200, message = "La raison sociale ne peut pas dépasser 200 caractères")
    private String raisonSociale;

    @Column(length = 20)
    @Size(max = 20, message = "L'IF ne peut pas dépasser 20 caractères")
    @Pattern(regexp = "^[0-9]*$", message = "L'IF doit contenir uniquement des chiffres")
    private String identifiantFiscal; // IF

    @Column(unique = true, length = 15)
    @Size(min = 15, max = 15, message = "L'ICE doit contenir exactement 15 caractères")
    @Pattern(regexp = "^[0-9]{15}$", message = "L'ICE doit contenir exactement 15 chiffres")
    private String ice;

    @Column(columnDefinition = "TEXT")
    private String adresse;

    @Column(length = 100)
    @Size(max = 100, message = "La ville ne peut pas dépasser 100 caractères")
    private String ville;

    @Column(length = 20)
    @Pattern(regexp = "^[0-9+\\-\\s()]*$", message = "Le téléphone doit contenir uniquement des chiffres et caractères spéciaux valides")
    private String telephone;

    @Column(length = 20)
    @Pattern(regexp = "^[0-9+\\-\\s()]*$", message = "Le fax doit contenir uniquement des chiffres et caractères spéciaux valides")
    private String fax;

    @Column(length = 150)
    @Email(message = "L'email doit être valide")
    private String email;

    @Column(length = 100)
    @Size(max = 100, message = "Le contact ne peut pas dépasser 100 caractères")
    private String contact;

    @Column(nullable = false)
    private Boolean nonResident = false; // Pour application IS tiers 10%

    @Column(columnDefinition = "TEXT")
    private String remarques;
}

File: backend/src/main/kotlin/ma/investpro/entity/Fournisseur.kt
package ma.investpro.entity

import jakarta.persistence.*
import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Pattern

@Entity
@Table(
    name = "fournisseurs",
    indexes = [
        Index(name = "idx_fournisseurs_code", columnList = "code"),
        Index(name = "idx_fournisseurs_ice", columnList = "ice"),
        Index(name = "idx_fournisseurs_actif", columnList = "actif"),
        Index(name = "idx_fournisseurs_non_resident", columnList = "non_resident")
    ]
)
class Fournisseur(
    @Column(nullable = false, unique = true, length = 50)
    @field:NotBlank
    var code: String = "",

    @Column(name = "raison_sociale", nullable = false, length = 200)
    @field:NotBlank
    var raisonSociale: String = "",

    @Column(name = "identifiant_fiscal", length = 20)
    @field:Pattern(regexp = "^[0-9]*$")
    var identifiantFiscal: String? = null,

    @Column(unique = true, length = 15)
    @field:Pattern(regexp = "^[0-9]{15}$")
    var ice: String? = null,

    @Column(columnDefinition = "TEXT")
    var adresse: String? = null,

    @Column(length = 100)
    var ville: String? = null,

    @Column(length = 20)
    var telephone: String? = null,

    @Column(length = 20)
    var fax: String? = null,

    @Column(length = 150)
    @field:Email
    var email: String? = null,

    @Column(length = 100)
    var contact: String? = null,

    @Column(name = "non_resident", nullable = false)
    var nonResident: Boolean = false,

    @Column(columnDefinition = "TEXT")
    var remarques: String? = null

) : BaseEntity()
